### PR TITLE
Add the class FrechetMean in learning

### PR DESCRIPTION
**Pull request by ninamiolane**
Start to address issue #364 

Create class FrechetMean in learning
Remove code for frechet mean in the RiemannianMetric class
Add three private functions for the current three methods for gradient descent
Address edge case for the frechet mean on linear spaces, Euclidean and Minkowski

### CODE REVIEW

File: geomstats/learning/frechet_mean.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+    var = gs.array(var)
+    var /= sum_weights
+
+    var = gs.to_ndarray(var, to_ndim=1)
+    var = gs.to_ndarray(var, to_ndim=2, axis=1)
+    return var
+
+
+def linear_mean(points, weights=None):
+    """Compute the frechet mean.
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: Maybe simpler to state this the other way around:
"""Compute the euclidean mean.

The Euclidean mean coincides with the Frechet mean when points lie in a Euclidean space with Euclidean metric.
"""

File: geomstats/learning/frechet_mean.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        points = gs.vstack(points)
+    points = gs.to_ndarray(points, to_ndim=2)
+    n_points = gs.shape(points)[0]
+
+    if isinstance(weights, list):
+        weights = gs.vstack(weights)
+    elif weights is None:
+        weights = gs.ones((n_points, ))
+
+    weighted_points = gs.einsum('n,nj->nj', weights, points)
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: use ellipsis for vectorized einsum?

File: geomstats/learning/frechet_mean.py [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+        return mean
+
+    sq_dists_between_iterates = []
+    iteration = 0
+    sq_dist = gs.array([[0.]])
+    var = gs.array([[0.]])
+
+    last_iteration, mean, var, sq_dist = gs.while_loop(
+        lambda i, m, v, sq: while_loop_cond(i, m, v, sq),
+        lambda i, m, v, sq: while_loop_body(i, m, v, sq),
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: How did you get lambda functions to pass flake8??

File: geomstats/learning/frechet_mean.py [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        loop_vars=[iteration, mean, var, sq_dist],
+        maximum_iterations=n_max_iterations)
+
+    if last_iteration == n_max_iterations:
+        print('Maximum number of iterations {} reached.'
+              'The mean may be inaccurate'.format(n_max_iterations))
+
+    if verbose:
+        print('n_iter: {}, final variance: {}, final dist: {}'.format(
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: `final dist` is ambiguous (especially after variance), I don't remember what scipy.optimize uses but maybe stg like `gap` or a mention that it is the distance between the last two iterates.

File: geomstats/learning/frechet_mean.py [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+    iteration = 0
+    convergence = math.inf
+    barycenter = points.mean(0, keepdims=True) * 0
+
+    while convergence > tau and n_max_iterations > iteration:
+
+        iteration += 1
+
+        expand_barycenter = gs.repeat(barycenter, points.shape[0], 0)
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: Is this necessary? I thought the properly vectorized log could take `n_samples` points with only one `base_point` (works for the sphere at least)

File: geomstats/learning/frechet_mean.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+        print('n_iter: {}, final variance: {}, final dist: {}'.format(
+            last_iteration, var, sq_dist))
+
+    mean = gs.to_ndarray(mean, to_ndim=2)
+    return mean
+
+
+def _ball_gradient_descent(points, metric, weights, n_max_iterations):
+    lr = 1e-3
+    tau = 5e-3
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: Shouldn't these be arguments of the method so that users can play with them?

File: geomstats/learning/frechet_mean.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+def _adaptive_gradient_descent(points,
+                               metric,
+                               weights=None,
+                               n_max_iterations=32,
+                               epsilon=1e-12,
+                               init_points=[]):
+    """Compute the Frechet mean using gradient descent.
+
+    Frechet mean of (weighted) points using adaptive time-steps
+    The loss function optimized is ||M_1(x)||_x (where M_1(x) is
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: use :math:

File: geomstats/learning/frechet_mean.py [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+
+    Parameters
+    ----------
+    points: array-like, shape=[n_samples, dimension]
+
+    weights: array-like, shape=[n_samples, 1], optional
+
+    init_points: array-like, shape=[n_init, dimension]
+
+    epsilon: tolerance for stopping the gradient descent
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: write this on two lines, with `epsilon : float` on the first

File: geomstats/learning/frechet_mean.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+    """
+    # TODO(xpennec): This function assumes that all points are lists
+    #  of vectors and not of matrices
```
Comment (nguigs) [inline geomstats/learning/frechet_mean.py]: maybe raise an error if `point_type=='matrix'`

File: geomstats/learning/frechet_mean.py (hunk 1 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: Good point, thanks. I just used "linear" instead of Euclidean, because this is valid for both the euclidean mean and the minkowski mean.

File: geomstats/learning/frechet_mean.py (hunk 2 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: YESSSS! I just checked, and tensorflow 2.0 now supports ellipsis! This is gonna make the vectorization so much easier 🎉

File: geomstats/learning/frechet_mean.py (hunk 3 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: haha, no idea!

File: geomstats/learning/frechet_mean.py (hunk 5 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: Good eye, I've deleted. Thanks.

File: geomstats/learning/frechet_mean.py (hunk 6 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: Yes, done.

File: geomstats/learning/frechet_mean.py (hunk 7 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: Done.

File: geomstats/learning/frechet_mean.py (hunk 8 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: Yes, done.

File: geomstats/learning/frechet_mean.py (hunk 9 above)
Comment (ninamiolane) [reply geomstats/learning/frechet_mean.py]: Yes, done!
